python/ray/tests/test_iter.py

```diff
@@ -158,6 +158,30 @@ def test_for_each(ray_start_regular_shared):
     assert list(it.gather_sync()) == [0, 4, 2, 6]
 
 
+def test_for_each_concur(ray_start_regular_shared):
+    from time import sleep, perf_counter
+
+    def task(x):
+        sleep(3)
+        return x * 2
+
+    it = from_range(8)
+    # Make sure there are enough warm workers
+    ray.get(
+        [ray.remote(task).options(num_cpus=0.1).remote(1) for _ in range(4)])
+
+    start = perf_counter()
+    it = it.for_each_concur(task, max_concur=2, resources={"num_cpus": 0.1})
+    result = list(it.gather_sync())
+    end = perf_counter()
+    elapsed = end - start
+
+    assert repr(
+        it) == "ParallelIterator[from_range[8, shards=2].for_each_concur()]"
+    assert result == [0, 8, 2, 10, 4, 12, 6, 14]
+    assert 6 < elapsed < 6.5
```

Comment:
This is almost guaranteed to be a flaky test, please refactor to not depend on timing. You can try using the `SignalActor` in `test_utils.py` to communicate between processes instead